GabijaBern/assignment2/BinaryTreeAncestors.java

```diff
@@ -0,0 +1,251 @@
+	}
+}
+/*
+ * Generic class for Nodes
+ * 
+ * Here we override toString that is used while testing the methods
+ */
+class Node<T> 
+{
+	public T value;
+	public BTree<T> left = null, right = null;
+	
+	public Node(T valueToSet)
+	{
+		this.value = valueToSet;
+	}
+	
+	@Override
+	public String toString()
+	{
+		return "" + value;
+	}
+}
+/*
+ * Binary tree implementation with insertion and ancestors methods
+ */
+class BTree<T>
+{
+	Node<T> treeRoot = null;
+	
+	/*
+	 * Method for filling levels of tree
+	 * we keep a queue of nodes that don't have two children yet, every time we add a node
+	 * we add it to the end of the queue and make it a child of a first node in a queue
+	 * once the node in front has to children, we remove the node from the list
+	 * 
+	 * @param tree The tree into which we want to insert a new node
+	 * @param valueToInsert The value that we want to insert
+	 * 
+	 * @throws NoSuchElementException Exception can be thrown if queue is empty
+	 */
+	public void insert(Node<T> tree, T valueToInsert) throws NoSuchElementException
+	{
+		Queue<Node<T>> nodesQueue = new LinkedList<Node<T>>();
+        nodesQueue.add(tree);
+ 
+        while (!nodesQueue.isEmpty()) 
+        {
+        	//we take first node that doesn't have two children in the list
+        	Node<T> firstNode = nodesQueue.peek();
+        	nodesQueue.remove();
+        	
+        	//we start filling from left to right
+            if (firstNode.left == null)
+            {
+            	firstNode.left = new BTree<T>();
+                firstNode.left.treeRoot = new Node<T>(valueToInsert);
+                break;
+            }
+            else
+            	nodesQueue.add(firstNode.left.treeRoot);
+            
+            //if left child is not empty, we try to fill the right node
+            if (firstNode.right == null)
+            {
+            	firstNode.right = new BTree<T>();
+                firstNode.right.treeRoot = new Node<T>(valueToInsert);
+                break;
+            }
+            else
+            	nodesQueue.add(firstNode.right.treeRoot);
+   
+        }
+	}
+
+	/*
+	 * Method for finding a list of all node ancestors
+	 * 
+	 * @param givenTree The tree in which we want to find all ancestors
+	 * @param nodeToFind The node whose ancestors we want to find in the tree
+	 * 
+	 * @return list including the node itself (first element in the list), the list is empty 
+	 * if node is not in the tree
+	 */
+	public boolean findAncestors(BTree<T> givenTree, Node<T> nodeToFind)
+	{
+		
+		//if we are given a null node or tree to search for/in
+		if(nodeToFind  == null || givenTree == null)
+			return false;
+		
+		//base case - we found the node we have been looking, so we add it to the list
+		if(givenTree.treeRoot.value == nodeToFind.value)
+		{	
+			System.out.print("Node: " + givenTree.treeRoot.value + " ancestors: ");

```

Comment:
It is good design to not print as you go along, but instead collect the results in some sort of data structure, then print at the very end. This makes it much easier to unit test. Could you try doing this?